Add Silicon Heaven protocol adaptation into NuttX and SHV examples

zdebanos opened the pull request:
This commit marks the end of my GSoC 2025 project in the NuttX section.

Currently creating as a draft pull request.
I have agreed with Pavel Pisa to separate the library's dependancy named ulut from the shv-libs4c repository (located currently here https://github.com/silicon-heaven/shv-libs4c/tree/master/submodule). This will require more time for me. Also, I have discussed with Karel Koci from Elektroline a.s. about the posibilities of the exports of external libraries - something that I somehow solved in this commit: https://github.com/apache/nuttx-apps/pull/3154.
In the end we have agreed that this should be considered as "hacking the whole Makefile system" and I should rather come up with a better solution, and thus I'm postponing it. Since I require the export for the GSoC, I'd be glad to leave it here as draft until I solve all the issues already mentioned.

I am about to add documentation into the nuttx repo, too.

## Summary

Please look into the commit's verbose message :-).

## Impact

Adds the Silicon Heaven protocol implementation. A framework designed to register your callbacks into your application. The callbacks are organized into nodes and the nodes can form a tree structure.

## Testing

The testing was done on my custom samv7 board. In nuttx:
`tools/configure.sh myboard/myconfig` - configuration OK
`make` - downloads the shv-libs4c repo from github, and compiles OK

DISCLAIMER: in my GSoC, I was using `make export`. But with the solution used here (https://github.com/zdebanos/pysimCoder/tree/shv-refactor) I succesfully compiled against automatically generated source code in pysimCoder.

michallenc commented on examples/shv-nxboot-updater/update-script/gui.py:
```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python3
```
add `requirements.txt` with packages needed to run those scripts

zdebanos replied:
Fixed

michallenc commented on examples/shv-nxboot-updater/update-script/shvconfirm.py:
```diff
@@ -0,0 +1,31 @@
+# under the License.
+#
+############################################################################
+from shv import RpcUrl, ValueClient
```
These imports won't work with `pyshv >= 0.10.0`. I don't think we need backwards compatibility here, just change imports and ensure `pyshv >= 0.10.0` in `requirements.txt `

zdebanos replied:
Fixed

michallenc commented on examples/shv-nxboot-updater/shv_nxboot_updater.c:
```diff
@@ -0,0 +1,424 @@
+#include <sys/ioctl.h>
+#include <semaphore.h>
+
+#ifdef __NuttX__
```
This is always true in NuttX apps.

zdebanos replied:
Fixed. A relict from previous experiments.

michallenc commented on examples/shv-nxboot-updater/shv_nxboot_updater.c:
```diff
@@ -0,0 +1,424 @@
+{
+  shv_unpack_data(&shv_ctx->unpack_ctx, 0, 0);
+  nxboot_confirm();
+  shv_send_int(shv_ctx, rid, 0);
```
```suggestion
  shv_send_empty_response(shv_ctx, rid);
```

zdebanos replied:
Fixed.

michallenc commented on examples/shv-nxboot-updater/shv_nxboot_updater.c:
```diff
@@ -0,0 +1,424 @@
+  .method = shv_fwstable_confirm
+};
+
+static const shv_method_des_t * const shv_dev_fwstable_dmap_items[] =
```
there should also be `fwStable:get` method that returns `true` or `false` based on fw status.

michallenc commented on examples/shv-nxboot-updater/shv_nxboot_updater.c:
```diff
@@ -0,0 +1,424 @@
+                             int rid)
+{
+  shv_unpack_data(&shv_ctx->unpack_ctx, 0, 0);
+  shv_send_int(shv_ctx, rid, 1);
```
This is SHV version major, therefore it should be `3` as we comply with version `3.0`

zdebanos replied:
I'm about to add a .app node into shv-libs4c library. After the fix, I shall remove the .app node definition.

michallenc commented on examples/shv-test/shv_test.c:
```diff
@@ -0,0 +1,510 @@
+                             int rid)
+{
+  shv_unpack_data(&shv_ctx->unpack_ctx, 0, 0);
+  shv_send_int(shv_ctx, rid, 1);
```
```suggestion
  shv_send_int(shv_ctx, rid, 3);
```

zdebanos replied:
I'm about to add a .app node into shv-libs4c library. After the fix, I shall remove the .app node definition.

michallenc commented on examples/shv-nxboot-updater/update-script/gui.py:
```diff
@@ -0,0 +1,221 @@
+        '-i', '--image',
+        dest='image',
+        type=str,
+        default='update.img',
```
```suggestion
        default='nuttx.nximg',
```

zdebanos replied:
Fixed.

michallenc commented on examples/shv-nxboot-updater/update-script/gui.py:
```diff
@@ -0,0 +1,221 @@
+        '-m', '--mount',
+        dest='target_mount',
+        type=str,
+        default='test/samocon-1',
```
Change to something nuttx generic

zdebanos replied:
Fixed.

michallenc commented on examples/shv-nxboot-updater/update-script/shvflasher.py:
```diff
@@ -0,0 +1,80 @@
+import io
+import zlib
+
+from shv import RpcUrl, SHVBytes, ValueClient
```
ditto

michallenc commented on examples/shv-nxboot-updater/shv_nxboot_updater.c:
```diff
@@ -0,0 +1,424 @@
+static const shv_method_des_t shv_dev_root_dmap_item_device_type =
+{
+  .name = "deviceType",
+  .method = shv_root_device_type
```
What is the purpose of this method?

michallenc commented on examples/shv-test/shv_test.c:
```diff
@@ -0,0 +1,510 @@
+{
+  shv_unpack_data(&shv_ctx->unpack_ctx, &g_testing_val, 0);
+  printf("Testing val set to %d\n", g_testing_val);
+  shv_send_int(shv_ctx, rid, 0);
```
```suggestion
  shv_send_empty_response(shv_ctx, rid);
```

zdebanos commented on netutils/libshvc/Makefile:
```diff
@@ -0,0 +1,60 @@
+	
+# Compile the library. Also pass all the NuttX Makefile variables.
+compile-libshv: $(SHVDIR)
+	- $(Q) ( cd $(SHVDIR) && git fetch origin dev-zdebanos ) > /dev/null 2>&1
```
Already removed this line.

zdebanos commented:
I came across an issue, which must be fixed: locally, I was running against `pyshv==0.7.3` (implementation of Silicon Heaven API in python) but when I added requirements and changed the imports to match with pyshv==0.10.0, it stopped working, probably by using a bad function of some sort. I'll have to look deeper into that.

All i know is that the script worked with the previous version, the target device functionality has not changed.

zdebanos commented:
@simbit18 I know about that, I'm working on a different fix and I'm gonna fix the linter error too.

zdebanos commented:
I have unmarked this commit as draft since we have agreed upon a stable solution. The only thing that needs to be done is the documention, I'm working on it right now.

ppisa commented on netutils/libshvc/Kconfig:
```diff
@@ -0,0 +1,11 @@
+
+config NETUTILS_LIBSHVC
+	bool "Silicon Heaven Protocol (lightweight C implementation)"
+	select NETUTILS_LIBULUT
```
I am not sure.  The depend would hide SHV support if LIBULUT is not enabled and we do not want to enable it by default. When somebody looks for SHV (which is probable start point/intention) then enable of SHV would enable ULUT when select is used. And ULUT is not used for anything else in the NuttX for now. It can change in future if we want and common sense would consider useful to include some other our project utilizing uLUt

* uLan communication “Open Source implementation of a multi-master RS485 communication protocol” https://ulan.sourceforge.net/ (already tested with NuttX and includes NuttX support in mainline in addition to Linux, Windows and OS system-less devices - sysless)
* OCERA Real-Time CAN https://ortcan.sourceforge.net/ with dynamic, EDS lead based CAN open device, manager and monitor (little dated unfortunately, but with extension to CAN FD interesting project still) (Linux NuttX and system less support included)
*  Open Real-Time Ethernet (ORTE) DDS RTPS C language implementation https://orte.sourceforge.net/ Again without care for long time and only protocol version 1.0 at that time. But it has been exactly the worldwide second implementation compatible with original RTI NDDS and that interoperability has been base for DDS Internet standard acceptance. It can be interesting still if upgrade to DDS RTPS 2.0 version is finished ([there is the start](https://wiki.control.fel.cvut.cz/mediawiki/images/3/37/Dp_2016_hubacek_jiri.pdf)). 
* Small User Interface Toolkit (SuiTk) https://gitlab.com/pikron/sw-base/suitk/-/wikis/home providing higher level layer above Microwindows which allows for hand written C code run on systemless MCUs with 64 kB RAM and still utilizing event connectors (Qt signal slot equivalents) and with little more RAM it can use full declarative interface, busyness logic description dynamically build from files with XML syntax subset.

So there is some potential for uLUt broader use with NuttX, but it is only support dependency for SHV in NuttX for now. When situation changes we can reconsider the choice.

raiden00pl replied:
`select XXX` has its problems and should be avoided when possible. There were discussions about this topic in several places on github lately (I don't remember exactly where) and on the mailing list: https://lists.apache.org/thread/n7qq5341cln5d3d4roy5nj23yc8hgzn6

The conclusion is that we should use `select` only for options that are not visible to the user. BTW, the user can always show hidden symbols using "show-all mode" in menuconfig.

raiden00pl replied:
We can use something like this:

```
if !NETUTILS_LIBULUT && !NETUTILS_LIBSHVC
comment "uLan Utilities Library requiered to support SHV"
endif

```

zdebanos replied:
I agree with @raiden00pl, you see what it's for but the moment you turn it on, it dissapears and it does not bloat the kconfig menu.

zdebanos replied:
@raiden00pl @ppisa fixed. Also I've tested that the whole system compiles with ulut only and ulut + shv.

ppisa commented on netutils/libshvc/Makefile:
```diff
@@ -0,0 +1,78 @@
+$(SHVOBJS) : shv-build-library
+
+$(SHVDIR):
+	$(call CLONE, https://github.com/silicon-heaven/shv-libs4c, $(SHVDIR), )
```
I am not sure what is the best option. If the repository is already cloned then there should be mechanism how to keep it at the correct version matching given NuttX version/commit.

I have implemented next script for integration into [pysimCoder](https://github.com/robertobucher/pysimCoder)

https://github.com/zdebanos/pysimCoder/blob/gsoc-2025-shv-refactor/ExtLibs/extlib-fetch.sh

It keeps the `local` branch, if the repository is clean it switches it to the specified library version by reset. If there is only small local change which does not clash with switch to newly required commit, it informs user and switches. If there are conflicting changes it fails. If there is another branch and state is clean it switches and it is user's task to move changes.

I am not sure what is preferred and typical for NuttX

zdebanos replied:
I've agreed with @ppisa that the solution with tar.gz is possible but as we do not have any release yet, we will use the commit hash instead as an identifier. We also both think a hash identifier is a more reasonable solution.

I'll try to submit a solution as fast as I can.

zdebanos replied:
@cederom @raiden00pl  I've fixed this and now stable tarballs are downloaded. We use the hash identifier to download the correct version. Therefore, I consider this fixed. As soon as you ACK this, I resolve the conversation.

zdebanos commented on netutils/libulut/Makefile:
```diff
@@ -0,0 +1,78 @@
+$(ULUTOBJS) : ulut-build-library
+
+$(ULUTDIR):
+	$(call CLONE, https://git.code.sf.net/p/ulan/ulut.git, $(ULUTDIR), )
```
Ditto
